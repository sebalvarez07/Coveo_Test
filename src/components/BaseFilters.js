import React from 'react';
import { connect } from 'react-redux';
import { startUpdateBaseFilter } from '../actions/filters';

export class BaseFilters extends React.Component {

    constructor (props) {
        super(props);
        this.beerButton = React.createRef();
        this.wineButton = React.createRef();
        this.state = {
            beerActive: true,
            wineActive: true
        }
    }

    componentDidUpdate (prevProps, prevState) {
        if(prevState.beerActive !== this.state.beerActive) {
            this.beerButton.current.classList.toggle('active-base');
        }
        
        if(prevState.wineActive !== this.state.wineActive) {
            this.wineButton.current.classList.toggle('active-base');
        }
    }
    
    dispatchBaseFilter = () => {
        let baseFilter;
        const beerActive = this.state.beerActive;
        const wineActive = this.state.wineActive;

        if(beerActive === wineActive) {
            baseFilter = ['Bière', 'Vin'];
        }
        else if(beerActive && !wineActive) {
            baseFilter = ['Bière'];
        }
        else if(!beerActive && wineActive) {
            baseFilter = ['Vin'];
        }

        this.props.startUpdateBaseFilter(baseFilter);
    }

    handleBaseFilter = (e) => {
        e.preventDefault(); 
        const btn = e.target.closest('.base-filter__btn');

        if(btn.id === 'beerFilter') {
            this.setState(prevState => (
                {
                    beerActive: !prevState.beerActive
                }
                ), this.dispatchBaseFilter);
        }
        else {
            this.setState(prevState => (
                {
                    wineActive: !prevState.wineActive
                }
                ), this.dispatchBaseFilter);
        }
    }

    render () {
        return (
            <div className='base-filter header__el'>
                <span 
                    ref={this.beerButton}
                    className='base-filter__btn active-base'
                    id={'beerFilter'}
                    onMouseDown={this.handleBaseFilter}
                >
                    <i className="ionicons ion-ios-pint"></i>
                    <span className='base-filter__label'>
                        Beer
                    </span>
                </span>
                <span
                    ref={this.wineButton}
                    className='base-filter__btn active-base'
                    id={'wineFilter'}
                    onMouseDown={this.handleBaseFilter}
                >
                    <i className="ionicons ion-ios-wineglass"></i>
                    <span className='base-filter__label'>
                        Wine
                    </span>
                </span>
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startUpdateBaseFilter: (baseFilter) => dispatch(startUpdateBaseFilter(baseFilter))
    }
}
 
export default connect(undefined, mapDispatchToProps)(BaseFilters);