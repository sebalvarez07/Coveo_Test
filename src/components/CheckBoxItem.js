import React, { useState } from 'react';

const CheckBoxItem = (props) => {

};

export default CheckBoxItem;