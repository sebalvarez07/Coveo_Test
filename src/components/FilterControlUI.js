import React from 'react';
import { connect } from 'react-redux';
import { groupMap } from '../fixtures/helpers';
import { startRemoveFilterValue } from '../actions/filters';

export const FilterControlUI = (props) => {

    const handleValueRemoval = (e) => {
        const btn = e.target.closest('.filter-control__btn');
        const group = btn.dataset.group;
        const val = btn.dataset.value;
        props.startRemoveFilterValue(group, val);
    } 

    return (
        <div className='filter-controlUI'>
            <h4 className='current-viewing'>
                Currently Viewing: {props.baseFilter.join(' & ')}
            </h4>
            {
                Object.keys(props.filtersActive).map(category => {
                    if(props.filtersActive[`${category}`].values.length > 0) {
                        return (
                            <div className='filter-control__container' key={category}>
                                <span className='filter-control__cat'>{groupMap.get(category)}: </span>
                                {
                                    props.filtersActive[category].values.map(activeFilterVal => {
                                        return <button 
                                                    data-group={category}
                                                    data-value={activeFilterVal}
                                                    className='filter-control__btn'
                                                    onMouseDown={handleValueRemoval}
                                                    key={activeFilterVal}>
                                                        {activeFilterVal}
                                                        <i className="ionicons ion-close"></i>
                                                </button>
                                    })
                                }
                            </div>
                        )    
                    }
                })
            }
        </div>
    )
}

const mapStateToProps = (state) => (
    {
        filtersActive: state.filters.categories,
        baseFilter: state.filters.baseFilter
    }
);

const mapDispatchToProps = (dispatch) => (
    {
        startRemoveFilterValue: (group, value) => dispatch(startRemoveFilterValue(group, value)),
    }
);

export default connect(mapStateToProps, mapDispatchToProps)(FilterControlUI);