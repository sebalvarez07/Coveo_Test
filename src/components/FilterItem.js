import React from 'react';
import { connect } from 'react-redux';
import { startUpdateCategory, startUpdateCategoryLimit } from'../actions/filters';
import { groupMap } from '../fixtures/helpers';

const defaultLimit = 5;

class FilterItem extends React.Component {

    constructor(props) {
        super(props);
        this.checkBoxes = [];
        this.checkedItems = new Set();
        this.state = {
            valueLimit: defaultLimit
        }
    }

    componentWillMount () {
        this.checkBoxes = this.props.filter.map(option => option.value);
    }

    componentWillUpdate (nextProps) {

        this.checkedItems = new Set([...this.props.checkedFilters]);

        if(nextProps.baseFilter !== this.props.baseFilter) {
            // TODO.. on baseFilter change, remove wine || beer filters           
        }
        
        // Ensure to keep check boxes on view
        this.checkBoxes = [...this.checkedItems];
    
        // Keep old checked boxes, and add new boxes
        nextProps.filter.forEach(option => {
            if(!this.checkedItems.has(option.value)){
                this.checkBoxes.push(option.value)
            }
        });     
    }

    handleChange = (e) => {
        const item = e.target.name;
        const isChecked = e.target.checked;
            
        if(isChecked) {
            this.checkedItems.add(item)
        } else {
            if(this.checkedItems.has(item)) {
                this.checkedItems.delete(item);
            }
        }

        const checkedItemsArr = Array.from(this.checkedItems);

        this.props.startUpdateCategory(this.props.filterName, checkedItemsArr);
    }

    dispatchMoreValues = () => {
        if(this.state.valueLimit <= this.props.filter.length) {
            return;
        } else {
            this.props.startUpdateCategoryLimit(this.props.filterName, this.state.valueLimit);
        }
    }

    handleValueLimit = (e) => {
        e.preventDefault();
        
        if(e.target.name === 'more'){
            this.setState(prevState => {
                return {
                    valueLimit: prevState.valueLimit + 5
                }
            }, this.dispatchMoreValues)
        }

        else if(e.target.name === 'less'){
            const resetLimit = this.checkedItems.size > defaultLimit ? this.checkedItems.size : defaultLimit;
            this.setState({ valueLimit: resetLimit});
        }
    }

    handleClearAll = () => {
        this.props.startUpdateCategory(this.props.filterName, []);
    }

    setLimit = () => {
        
        return this.checkedItems.size > this.state.valueLimit ? this.checkedItems.size : this.state.valueLimit;
    }

    renderCheckBox = (checkName) => {        
        return (
            <label className='check__container' key={`${checkName}`}>
                {`${checkName}`}
                <input 
                    className='custom__checkbox'
                    style={{ display: 'block' }} 
                    type='checkbox' 
                    name={`${checkName}`} 
                    checked={this.checkedItems.has(checkName)}
                    onChange={this.handleChange} 
                />
                <span className='checkmark'>
                    <i className="ionicons ion-checkmark"></i>
                </span>
            </label>
        )
    }
    
    render (){
        const limit = this.setLimit();
        return (
            <div className='sidebar__group'>
                <div className='group__header'>
                    <h3 className='group__tl'>{groupMap.get(this.props.filterName)}</h3>
                    <span className='clear-btn' onMouseDown={this.handleClearAll}>Clear All</span>
                </div>
                
                {
                    this.checkBoxes.reduce((acc, checkName, index) => {
                        if(limit > index) {
                            acc.push(this.renderCheckBox(checkName))
                        }
                        return acc;
                    }, [])
                }
                
                <div className='more_less__group'>
                    { this.state.valueLimit <= this.props.filter.length && 
                        <React.Fragment>
                            <button 
                                className='more_less__btn'
                                name={'more'}
                                onMouseDown={this.handleValueLimit}
                                >
                                Show More
                            </button>
                            <span> | </span>
                        </React.Fragment>
                    }
                    { this.props.filter.length >= defaultLimit && 
                        <button 
                            className='more_less__btn'
                            name={'less'}
                            onMouseDown={this.handleValueLimit}
                            >
                            Show Less
                        </button>
                    }
                </div>
                
            </div>
        )
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startUpdateCategory: (group, update) => dispatch(startUpdateCategory(group, update)),
        startUpdateCategoryLimit: (group, valueLimit) => dispatch(startUpdateCategoryLimit(group, valueLimit))
    }
}

const mapStateToProps = (state, props) => {
    return  {
        baseFilter: state.filters.baseFilter,
        checkedFilters: state.filters.categories[`${props.filterName}`]['values']
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(FilterItem);