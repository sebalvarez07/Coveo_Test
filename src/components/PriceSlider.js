import React from 'react';
import { connect } from 'react-redux';
import { startUpdateCategory } from '../actions/filters';
import SliderComponent from './Slider';

const defaultValues = [0, 300];

export class PriceSlider extends React.Component {

    constructor(props) {
      super(props);
    }

    state = {
      domain: [0, 300],
      values: defaultValues.slice(),
      update: defaultValues.slice()
    }

    componentWillUpdate(nextProps, nextState) {
      if(nextProps.priceFilterVal !== this.props.priceFilterVal && nextProps.priceFilterVal.length === 0) {
          this.setState({
            values: defaultValues
          })           
      }
    }

    onUpdate = update => {
      this.setState({ update })
    }
  
    onChange = values => {
      this.setState({ values })
      const valOne = values[0] < values[1] ? values[0] : values[1];
      const valTwo = values[0] < values[1] ? values[1] : values[0];
  
      if(!(valOne === 0 && valTwo === 300)) {
        this.props.startUpdateCategory('tpprixnum', [`${valOne}..${valTwo}`]); 
      } else {
        this.props.startUpdateCategory('tpprixnum', []);
      }
    }

    render() {
        return (
            <div className='price-slider header__el'>
                <span className='price-slider__tl'>
                    Price Range
                </span>
                <SliderComponent 
                    domain={this.state.domain}
                    values={this.state.values}
                    update={this.state.update}
                    onChange={this.onChange}
                    onUpdate={this.onUpdate}
                />
                <span className='price-slider__current-range'>
                    {`${this.state.values[0]} $ - ${this.state.values[1]} $`}
                </span>
            </div>
        )
    }
}

const mapStateToProps = (state) => {
  return {
    baseFilter: state.filters.baseFilter,
    priceFilterVal: state.filters.categories.tpprixnum.values,
  }
}

const mapDispatchToProps = (dispatch) => {
    return {
        startUpdateCategory: (group, update) => dispatch(startUpdateCategory(group, update)),
    }
}
 
export default connect(mapStateToProps, mapDispatchToProps)(PriceSlider);