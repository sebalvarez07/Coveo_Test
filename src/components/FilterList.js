import React from 'react';
import { connect } from 'react-redux';
import FilterItem from './FilterItem';
 
export class FilterList extends React.Component {
    constructor(props) {
        super(props)
    }

    render () {
        return (
            <div>
                {
                    this.props.filters.map(filter => {
                        return (
                            <FilterItem key={`${filter.field}`} filterName={filter.field} filter={filter.values}/>
                        )
                    })
                }
            </div>
        )
    }
 }

 const mapStateToProps = (state) => (
    {
        filters: state.content.groupBy
    }
);

 export default connect(mapStateToProps)(FilterList);