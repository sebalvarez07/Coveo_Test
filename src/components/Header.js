import React from 'react';
import { connect } from 'react-redux';
import BaseFilters from './BaseFilters';
import SearchBar from './SearchBar';
import PriceSlider from './PriceSlider';

export class Header extends React.Component {

    constructor (props) {
        super(props);
    }

    render () {
        return (
            <header className="header">
                <div className="header-container">
                    <div className="logo-container">
                        <span className='logo'>coveo</span>
                    </div>
                    <div className="header__content">
                        <BaseFilters />
                        <PriceSlider />
                        <SearchBar />
                    </div>
                </div>
            </header>
        );
    }
}

export default connect()(Header);