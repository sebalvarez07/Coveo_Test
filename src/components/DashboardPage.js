import React from 'react';
import Header from './Header';
import Sidebar from './Sidebar';
import MainContainer from './MainContainer';
 
const DashboardPage = (props) => {
    return (
        <div className='wrapper'>
            <Header />
            <Sidebar />
            <MainContainer />
        </div>
    )
}

export default DashboardPage;