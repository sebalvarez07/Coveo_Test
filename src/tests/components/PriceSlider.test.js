import React from 'react';
import { shallow } from 'enzyme';
import { PriceSlider } from '../../components/PriceSlider';

let wrapper, startUpdateCategory;

beforeEach( () => {
    startUpdateCategory = jest.fn();
    wrapper = shallow(<PriceSlider startUpdateCategory={startUpdateCategory} />)
});

test('Should render base filters', () => {
    expect(wrapper).toMatchSnapshot();
}); 

test('Should handle startUpdateCategory when slider is changed', () => {
    // simulate mouseDown and pass the fake element's object 
    const valuesOG = [20, 100];
    const valuesProcessed = `${valuesOG[0]}..${valuesOG[1]}`;
    const filterName = 'tpprixnum';

    wrapper.find('SliderComponent').prop('onChange')(valuesOG);

    expect(startUpdateCategory).toHaveBeenLastCalledWith(filterName, [valuesProcessed]);
});