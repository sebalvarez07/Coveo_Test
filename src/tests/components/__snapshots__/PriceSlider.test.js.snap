// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render base filters 1`] = `
<div
  className="price-slider header__el"
>
  <span
    className="price-slider__tl"
  >
    Price Range
  </span>
  <SliderComponent
    domain={
      Array [
        0,
        300,
      ]
    }
    onChange={[Function]}
    onUpdate={[Function]}
    update={
      Array [
        0,
        300,
      ]
    }
    values={
      Array [
        0,
        300,
      ]
    }
  />
  <span
    className="price-slider__current-range"
  >
    0 $ - 300 $
  </span>
</div>
`;
